python/ccf/ledger.py

```
@@ -149,6 +164,200 @@ def _byte_read_safe(file, num_of_bytes):
+                cert = b"".join(self.node_certificates[signing_node])
+                existing_root = b"".join(values[self.EXPECTED_ROOT_INDEX])
+                signature = values[self.EXPECTED_NODE_SIGNATURE_INDEX][
+                    self.EXPECTED_SIGNATURE_INDEX
+                ]
+
+                tx_info = TxBundleInfo(
+                    self.merkle,
+                    existing_root,
+                    cert,
+                    signature,
+                    self.node_activity_status,
+                    signing_node,
+                )
+
+                # validations for 1, 2 and 3
+                # throws if ledger validation failed.
+                if self._verify_tx_set(tx_info):
+                    continue
+                else:
+                    raise ValueError("Ledger validation failed")
+
+        # Checks complete, add this transaction to tree
+        self.merkle.add_leaf(transaction.get_public_tx())
+
+    def verify_ending(self):
+        """
+        Verify the ledger ends in a signature transaction and has at least 1 signature.
+        These two conditions are prerequisite for the creation of .committed files.
+        This verifies the 4th condition.
+        """
+        if self.signature_count == 0:
+            LOG.error(
+                "Found 0 signatures. This usually means that the ledger is invalid"
+            )
+            raise SignatureCountException
+        if not self.signature_ending:
+            LOG.error(
+                "The last transaction in the ledger chunk wasn't a signature transaction which is unexpected."
+            )
+            raise SignatureEndingException
+
+    def _verify_tx_set(self, tx_info: TxBundleInfo) -> bool:
+        """Verify items 1, 2, and 3 for a the transactions up until a signature"""
+        # 1) The merkle root is signed by a TRUSTED node in the given network
+        if self._verify_node_status(tx_info):
```

Rather than returning a boolean, throwing a specific exception or at least logging which condition failed would probably be a good thing.

---

Commented too soon, but this still looks odd, because it suggests the calls can return False, which they happen never to do. If all errors are represented by exceptions (which I think is right here), then it seems to me the return status should be None.

---

Sounds good. Changed the signature of this and all the methods being called to return void and throw respective exceptions when verification failed.